docs/design/ha_global_restarts.md

```diff
@@ -0,0 +1,96 @@
+Every operator will add the processes that needs to be restarted under a special key e.g. `$lockKeyPrefix/pendingRestart/$processGroupID`.
+In this example the `$lockKeyPrefix` and `$processGroupID` will be replaced with the actual values during runtime.
+This ensures that all operators are aware of which processes must be restarted.
+Those changes alone are not enough to ensure that the coordination between the operators work.
+We have to signal all operators which processes have the correct configuration and can be restarted.
+To solve this each operator will add all process that are ready to be restarted under `$lockKeyPrefix/readyRestart/$processGroupID`.
+A process will be ready to be restarted if the current desired configuration is available, this will be validated by the operator (this step is already implemented in the current behaviour).
+This ensures that an operator only triggers the restart once all processes are ready to be restarted.
+Before doing the restart the operator must try to get a lock to ensure only one operator is restarting the processes.
+
+`$lockKeyPrefix/pendingRestart/$processGroupID` will contain all processes that meet the following conditions:
+
+- `IncorrectCommandLine` is set to true.
+- `SidecarUnreachable` is set to false.
+- `MissingProcesses` is set to false.
+- And the process group is not marked for removal.
+
+`$lockKeyPrefix/readyRestart/$processGroupID` will contain all processes that meet the following conditions:
+
+- `IncorrectCommandLine` is set to true.
+- `IncorrectPodSpec` is set to false.
+- `SidecarUnreachable` is set to false.
+- `MissingProcesses` is set to false.
+- And the process group is not marked for removal.
+
+The operator logic would change to add those keys when locking is enabled and the cluster is not being upgraded.
+The restart trigger logic would change like this:
+
+```go
+// Get all processes that have to be restarted, this method will return (map[string]struct{}, error)
+pendingRestart, err := readPendingRestart()
+...
+// Get all processes that are ready to restart, this method will return (map[string]struct{}, error)
+readyRestart, err := readReadyRestart(), this method will return (map[string]struct{}, error)
+...
+
+// We have to ensure that we only compare processes that are currently part of the cluster to ensure we ignore processes
+// that were deleted between the restart request.
+for _, process := range databaseStatus.Cluster.Processes {
+    processID := process.Locality[fdbv1beta2.FDBLocalityInstanceIDKey]
+
+    _, hasPendingRestart := pendingRestart[processID]
+    _, readyRestart := pendingRestart[processID]
+	
+	if hasPendingRestart != readyRestart {
+	    // wait until this process is ready for restart
+		return ...
+    }
+}
+
+// Ensure we clean the pending and the ready list before restarting the processes to ensure we have an empty list after the restart.
+err := clearPendingRestart()
+err := clearReadReadyRestart()
```

Comment:
Another approach would be to let the operator always update the key ranges `$lockKeyPrefix/pendingRestart` and `$lockKeyPrefix/readyRestart`. I decided against this to reduce the load on FDB. In the case that the operator Pod crashes during the clear steps and the kill command we could run into the risk that we are not restarting all processes at once (since every operator instance has to add the pending and ready information).